benchmarks/typescript.js

```diff
@@ -0,0 +1,82 @@
+'use strict';
+
+const { exec } = require('child_process');
+const mongoose = require('../');
+
+const numIterations = 10;
+
+let instantiations = 0;
+let memoryUsed = 0;
+let checkTime = 0;
+let totalTime = 0;
+
+run().catch(err => {
+  console.error(err);
+  process.exit(-1);
+});
+
+async function run() {
+  for (let i = 0; i < numIterations; ++i) {
+    console.log(`${i}...`);
+    const { stdout } = await new Promise((resolve, reject) => {
+      exec('../../../node_modules/.bin/tsc --extendedDiagnostics', { cwd: `${__dirname}/typescript/simple` }, (err, stdout, stderr) => {
+        if (err) {
+          console.error(err);
+          console.error(stdout);
+          console.error(stderr);
+          return reject(err);
+        }
+
+        resolve({ stdout });
+      });
+    });
+
+    const lines = stdout.split('\n');
+
+    const _instantiations = +lines.find(line => line.startsWith('Instantiations:')).match(/\d+/)[0];
+    const _memoryUsed = +lines.find(line => line.startsWith('Memory used:')).match(/\d+/)[0];
+    const _checkTime = +lines.find(line => line.startsWith('Check time:')).match(/\d+(\.\d+)?/)[0];
+    const _totalTime = +lines.find(line => line.startsWith('Total time:')).match(/\d+(\.\d+)?/)[0];
+
+    instantiations += _instantiations;
+    memoryUsed += _memoryUsed;
+    checkTime += _checkTime;
+    totalTime += _totalTime;
+  }
+
+  instantiations /= numIterations;
+  memoryUsed /= numIterations;
+  checkTime /= numIterations;
+  totalTime /= numIterations;
+  console.log(instantiations);
+  console.log(memoryUsed);
+  console.log(checkTime);
+  console.log(totalTime);
+
+  await persist({ instantiations, memoryUsed, checkTime, totalTime });
+}
+
+async function persist(results) {
+  if (!process.env.DB_URL) {
+    return;
+  }
+  if (!process.env.GITHUB_SHA) {
+    return;
+  }
+
+  await mongoose.connect(process.env.DB_URL);
+
+  const BenchmarkResult = mongoose.model('BenchmarkResult', mongoose.Schema({
+    githash: { type: String, required: true },
+    benchmarkName: { type: String, required: true },
+    results: 'Mixed'
+  }, { timestamps: true }), 'BenchmarkResult');
```

Comment:
We'd need to add an index to githash, I imagine this collection will grow over time.
```suggestion
    const benchmarkResultSchema = mongoose.Schema({
       githash: { type: String, required: true },
       benchmarkName: { type: String, required: true },
       results: 'Mixed'
     }, { timestamps: true });
     
     benchmarkResultSchema.index({ githash: 1 }, { unique: true });
     
    const BenchmarkResult = mongoose.model('BenchmarkResult', benchmarkResultSchema);
    await BenchmarkResult.syncIndexes();
```

Nitpick: Naming a collection `BenchmarkResult` conflicts with the default (therefore standard in my head) behavior of all lowercase, plural, collections names. I let mongoose do the pluralization instead.

Reply:
TBH, I prefer keeping the collection name the same as the model name. I personally have a strong aversion to unnecessary translation layers, so in our existing Mongoose backend logic we keep collection name = model name to avoid having to mentally calculate what Monngoose pluralizes "BenchmarkResult" to.

Reply:
As long as we're being consistent. 🤷
Even though I agree with your argument right now, since it's the default behavior and almost all projects use the default behavior, I'd rely on that.

But if someone were to be consistent with any pattern, it should be fine.